dev/build-tool-support/src/main/scala/com/lightbend/lagom/dev/Servers.scala

```diff
@@ -140,23 +142,33 @@ private[lagom] object Servers {
 
   private[lagom] object CassandraServer extends ServerContainer {
     protected type Server = {
-      def start(cassandraDirectory: File, configResource: String, clean: Boolean, port: Int, jvmOptions: Array[String]): Unit
+      def start(cassandraDirectory: File, yamlConfig: Option[File], clean: Boolean, port: Int, jvmOptions: Array[String]): Unit
       def stop(): Unit
       def address: String
       def hostname: String
       def port: Int
     }
 
-    def start(log: LoggerProxy, parentClassLoader: ClassLoader, classpath: Seq[File], port: Int, cleanOnStart: Boolean, jvmOptions: Seq[String], maxWaiting: FiniteDuration): Closeable = synchronized {
+    def start(
+      log:          LoggerProxy,
+      classpath:    Seq[File],
+      port:         Int,
+      cleanOnStart: Boolean,
+      jvmOptions:   Seq[String],
+      yamlConfig:   Option[File],
+      maxWaiting:   FiniteDuration
+    ): Closeable = synchronized {
+
       if (server != null) {
         log.info(s"Cassandra is running at ${server.address}")
       } else {
-        val loader = new java.net.URLClassLoader(classpath.map(_.toURI.toURL).toArray, parentClassLoader)
+
+        val loader = new java.net.URLClassLoader(classpath.map(_.toURI.toURL).toArray, this.getClass.getClassLoader)
```

Comment:
Here's what I've observed when running `sbt runAll`

On master, here's what `classpath` contains (with directories trimmed):

```
List(
  scala-library-2.11.12.jar,
  lagom-cassandra-server_2.11-1.4.0-RC1.jar,
  akka-persistence-cassandra-launcher_2.11-0.59.jar,
  akka-persistence-cassandra_2.11-0.59.jar,
  cassandra-driver-core-3.2.0.jar,
  netty-handler-4.0.44.Final.jar,
  netty-buffer-4.0.44.Final.jar,
  netty-common-4.0.44.Final.jar,
  netty-transport-4.0.44.Final.jar,
  netty-codec-4.0.44.Final.jar,
  guava-19.0.jar,
  metrics-core-3.1.2.jar,
  slf4j-api-1.7.7.jar,
  jnr-ffi-2.0.7.jar,
  jffi-1.2.10.jar,
  jffi-1.2.10-native.jar,
  asm-5.0.3.jar,
  asm-commons-5.0.3.jar,
  asm-tree-5.0.3.jar,
  asm-analysis-5.0.3.jar,
  asm-util-5.0.3.jar,
  jnr-x86asm-1.0.2.jar,
  jnr-posix-3.0.27.jar,
  jnr-constants-0.9.0.jar,
  akka-persistence_2.11-2.5.6.jar,
  akka-actor_2.11-2.5.6.jar,
  config-1.3.1.jar,
  scala-java8-compat_2.11-0.7.0.jar,
  akka-protobuf_2.11-2.5.6.jar,
  akka-cluster-tools_2.11-2.5.6.jar,
  akka-cluster_2.11-2.5.6.jar,
  akka-remote_2.11-2.5.6.jar,
  akka-stream_2.11-2.5.6.jar,
  reactive-streams-1.0.1.jar,
  ssl-config-core_2.11-0.2.2.jar,
  scala-parser-combinators_2.11-1.0.4.jar,
  netty-3.10.6.Final.jar,
  aeron-driver-1.3.0.jar,
  aeron-client-1.3.0.jar,
  agrona-0.9.6.jar,
  akka-persistence-query_2.11-2.5.6.jar
)
```

and here's what `parentClassLoader` contains:

```
[
  scala-library-2.11.12.jar,
  scala-compiler-2.11.12.jar,
  scala-reflect-2.11.12.jar,
  scala-xml_2.11-1.0.5.jar,
  scala-parser-combinators_2.11-1.0.4.jar,
  jline-2.14.3.jar
]
```

I looked up the chain of parents, and there isn't anything interesting before it gets to the system classloader.

On this pull request branch, here's `classpath`:

```
ArraySeq(
  scala-library-2.11.12.jar,
  lagom-cassandra-server_2.11.jar,
  akka-persistence-cassandra-launcher_2.11-0.59.jar,
  akka-persistence-cassandra_2.11-0.59.jar,
  cassandra-driver-core-3.2.0.jar,
  netty-handler-4.0.44.Final.jar,
  netty-buffer-4.0.44.Final.jar,
  netty-common-4.0.44.Final.jar,
  netty-transport-4.0.44.Final.jar,
  netty-codec-4.0.44.Final.jar,
  guava-19.0.jar,
  metrics-core-3.1.2.jar,
  slf4j-api-1.7.7.jar,
  jnr-ffi-2.0.7.jar,
  jffi-1.2.10.jar,
  jffi-1.2.10-native.jar,
  asm-5.0.3.jar,
  asm-commons-5.0.3.jar,
  asm-tree-5.0.3.jar,
  asm-analysis-5.0.3.jar,
  asm-util-5.0.3.jar,
  jnr-x86asm-1.0.2.jar,
  jnr-posix-3.0.27.jar,
  jnr-constants-0.9.0.jar,
  akka-persistence_2.11-2.5.6.jar,
  akka-actor_2.11-2.5.6.jar,
  config-1.3.1.jar,
  scala-java8-compat_2.11-0.7.0.jar,
  akka-protobuf_2.11-2.5.6.jar,
  akka-cluster-tools_2.11-2.5.6.jar,
  akka-cluster_2.11-2.5.6.jar,
  akka-remote_2.11-2.5.6.jar,
  akka-stream_2.11-2.5.6.jar,
  reactive-streams-1.0.1.jar,
  ssl-config-core_2.11-0.2.2.jar,
  scala-parser-combinators_2.11-1.0.4.jar,
  netty-3.10.6.Final.jar,
  aeron-driver-1.3.0.jar,
  aeron-client-1.3.0.jar,
  agrona-0.9.6.jar,
  akka-persistence-query_2.11-2.5.6.jar
)
```

(essentially identical)

Here's what's in `this.getClass.getClassLoader`:

```
[
  classes,
  lagom-sbt-plugin.jar,
  lagom-build-tool-support.jar,
  scala-library.jar,
  play-file-watch_2.10-1.1.7.jar,
  directory-watcher-0.3.2.jar,
  jna-4.2.1.jar,
  guava-23.0.jar,
  jsr305-1.3.9.jar,
  error_prone_annotations-2.0.18.jar,
  j2objc-annotations-1.1.jar,
  animal-sniffer-annotations-1.14.jar,
  slf4j-api-1.7.25.jar,
  better-files_2.10-2.17.0.jar,
  play-exceptions-2.6.10.jar,
  build-link-2.6.10.jar,
  cassandra-driver-core-3.2.0.jar,
  netty-handler-4.0.44.Final.jar,
  netty-buffer-4.0.44.Final.jar,
  netty-common-4.0.44.Final.jar,
  netty-transport-4.0.44.Final.jar,
  netty-codec-4.0.44.Final.jar,
  metrics-core-3.1.2.jar,
  jnr-ffi-2.0.7.jar,
  jffi-1.2.10.jar,
  jffi-1.2.10-native.jar,
  asm-5.0.3.jar,
  asm-commons-5.0.3.jar,
  asm-tree-5.0.3.jar,
  asm-analysis-5.0.3.jar,
  asm-util-5.0.3.jar,
  jnr-x86asm-1.0.2.jar,
  jnr-posix-3.0.27.jar,
  jnr-constants-0.9.0.jar,
  slf4j-nop-1.7.14.jar,
  sbt-plugin-2.6.10.jar,
  sbt-routes-compiler_2.10-2.6.10.jar,
  twirl-api_2.10-1.3.12.jar,
  commons-io-2.5.jar,
  run-support_2.10-2.6.10.jar,
  config-1.3.1.jar,
  sbt-twirl-1.3.12.jar,
  twirl-compiler_2.10-1.3.12.jar,
  scala-compiler.jar,
  scala-reflect.jar,
  twirl-parser_2.10-1.3.12.jar,
  sbt-native-packager-1.3.2.jar,
  commons-compress-1.14.jar,
  ant-1.10.1.jar,
  ant-launcher-1.10.1.jar,
  sbt-javaagent-0.1.4.jar,
  sbt-web-1.4.3.jar,
  akka-actor_2.10-2.3.16.jar,
  webjars-locator-core-0.33.jar,
  commons-lang3-3.1.jar,
  jackson-core-2.7.3.jar,
  sbt-js-engine-1.2.2.jar,
  jse_2.10-1.2.3.jar,
  akka-contrib_2.10-2.3.16.jar,
  akka-remote_2.10-2.3.16.jar,
  netty-3.8.0.Final.jar,
  protobuf-java-2.5.0.jar,
  uncommons-maths-1.2.2a.jar,
  akka-cluster_2.10-2.3.16.jar,
  akka-persistence-experimental_2.10-2.3.16.jar,
  leveldb-0.5.jar,
  leveldb-api-0.5.jar,
  leveldbjni-all-1.7.jar,
  leveldbjni-1.7.jar,
  hawtjni-runtime-1.8.jar,
  leveldbjni-osx-1.5.jar,
  leveldbjni-linux32-1.5.jar,
  leveldbjni-linux64-1.5.jar,
  leveldbjni-win32-1.5.jar,
  leveldbjni-win64-1.5.jar,
  trireme-core-0.8.9.jar,
  trireme-kernel-0.8.9.jar,
  rhino-1.7.7.1.jar,
  trireme-node10src-0.8.9.jar,
  spray-json_2.10-1.3.3.jar,
  npm_2.10-1.2.1.jar,
  npm-4.2.0.jar,
  sbteclipse-plugin-5.1.0.jar,
  scalaz-core_2.10-7.2.5.jar,
  scalaz-effect_2.10-7.2.5.jar
]
```

I didn't bother tracing up the line of parents from there.

So this works, but only by loading this with the sbt plugin's loader as the parent, which contains Scala 2.10 artifacts and a bunch of other things we don't need. I'm not sure it's such a good idea to have a classloader that mixes Scala 2.10 and 2.11 artifacts. It seems to work OK now, but it could be risky.

Reply:
I also tried modifying it to use _no_ parent class loader.

This causes an error:

```
java.lang.NoSuchMethodError: scala.Predef$.ArrowAssoc(Ljava/lang/Object;)Ljava/lang/Object;
```

I'm not sure why that happens...

I'm thinking that the previous existing class loader hierarchy with conversion to/from `null`. is the best solution on balance. It avoids making mixed up class paths and restricts the reflectively-invoked code to exactly what we expect.

Dealing with the `null` on the boundaries is not so bad:

```
server.start(directory, yamlConfig.orNull, cleanOnStart, port, jvmOptions.toArray)
```

```
val fileOrResource = Either.cond(yamlConfig == null, devEmbeddedYaml, yamlConfig)
```

Of course we should comment this code clearly so it will make sense to our future selves and anyone else reading the code.

I think this can all go away once we eventually remove support for sbt 0.13 and Scala 2.10 (assuming that sbt will keep up with new Scala versions in the future, as I believe is now planned).

What do you think? Can you live with this? :smile: